lib/CodeGen/CGDynamicCheck.cpp

```diff
@@ -139,80 +150,138 @@ void CodeGenFunction::EmitDynamicBoundsCheck(const Address BaseAddr,
+  // Emit the code to generate pointers for SubRange, lb and ub
   Address Lower = EmitPointerWithAlignment(SubRange->getLowerExpr());
   Address Upper = EmitPointerWithAlignment(SubRange->getUpperExpr());
 
-  Value *LowerInt = Builder.CreatePtrToInt(Lower.getPointer(), IntPtrTy,
-                                           "_Dynamic_check.lower");
-  Value *UpperInt = Builder.CreatePtrToInt(Upper.getPointer(), IntPtrTy,
-                                           "_Dynamic_check.upper");
+  // Emit the code to generate pointers for CastRange, castlb and castub
 
   Address CastLower = EmitPointerWithAlignment(CastRange->getLowerExpr());
-  Address CastUpper = EmitPointerWithAlignment(CastRange->getUpperExpr());
+  // We will be comparing CastLower to Lower. Their types may not match,
+  // so we're going to bitcast CastLower to match the type of Lower if needed.
+  if (CastLower.getType() != Lower.getType())
+    CastLower = Builder.CreateBitCast(CastLower, Lower.getType());
 
-  Value *CastLowerInt = Builder.CreatePtrToInt(CastLower.getPointer(), IntPtrTy,
-                                               "_Dynamic_check.castlower");
-  Value *CastUpperInt = Builder.CreatePtrToInt(CastUpper.getPointer(), IntPtrTy,
-                                               "_Dynamic_check.castupper");
+  Address CastUpper = EmitPointerWithAlignment(CastRange->getUpperExpr());
+  // Again we're going to bitcast CastUpper to match the type of Upper
+  // if needed.
+  if (CastUpper.getType() != Upper.getType())
+    CastUpper = Builder.CreateBitCast(CastUpper, Upper.getType());
 
   // Make the lower check (Lower <= CastLower)
-  Value *LowerChk =
-      Builder.CreateICmpULE(LowerInt, CastLowerInt, "_Dynamic_check.lower_cmp");
+  Value *LowerChk = Builder.CreateICmpULE(
+      Lower.getPointer(), CastLower.getPointer(), "_Dynamic_check.lower");
 
   // Make the upper check (CastUpper <= Upper)
-  Value *UpperChk =
-      Builder.CreateICmpULE(CastUpperInt, UpperInt, "_Dynamic_check.upper_cmp");
+  Value *UpperChk = Builder.CreateICmpULE(
+      CastUpper.getPointer(), Upper.getPointer(), "_Dynamic_check.upper");
 
-  Value *Cond2 = Builder.CreateAnd(LowerChk, UpperChk, "_Dynamic_check.range");
-  Builder.CreateCondBr(Cond2, DyCkSuccess, DyCkFail);
+  // Make Both Checks
+  Value *CastCond =
+      Builder.CreateAnd(LowerChk, UpperChk, "_Dynamic_check.cast");
 
-  Builder.SetInsertPoint(DyCkFail);
-  CallInst *TrapCall = Builder.CreateCall(CGM.getIntrinsic(Intrinsic::trap));
-  TrapCall->setDoesNotReturn();
-  TrapCall->setDoesNotThrow();
-  Builder.CreateUnreachable();
+  // Constant Folding:
+  // If CastCond is true (one), then we need to insert a direct branch
+  // to the success block, emit it, and set the insert point there for
+  // further code generation.
+  if (const ConstantInt *CastCondConstant = dyn_cast<ConstantInt>(CastCond)) {
+    if (CastCondConstant->isOne()) {
+      ++NumDynamicChecksElided;
 
```

Comment:
You asked "how do we know constant folding is correct". Well, turns out in writing comments about how it should work made me find a pretty major bug in it, which I have fixed in this most recent commit. 

I'll file a ticket about writing tests for this, but maybe in the short term I should also add a flag to tell the codegen to constant fold that defaults to `false`. What's your feeling on this?